keps/sig-api-machinery/0034-20190129-kflag.md

```diff
@@ -0,0 +1,188 @@
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Create a kflag package, which wraps pflag and addresses certain pain-points of dealing with flags
+in core Kubernetes components.
+For example: implementing componentconfig in a backwards-compatible way, or preventing a third-party
+library's flags from appearing in a component's command line interface.
+
+
+## Motivation
+
+Make it easier for component owners to implement
+[Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit#)
+(a.k.a. componentconfig).
+
+While a number of problems with existing machinery are highlighted in the above doc, this
+proposal is focused on the following two use-cases:
+* Enforcing [flag precedence](https://github.com/kubernetes/kubernetes/issues/56171) between values
+  specified on the command line and values from a config file, which is required to incrementally
+  migrate flags to a config file with backwards compatibility.
+* Preventing third-party code from
+  [implicitly registering flags](https://github.com/kubernetes/kubernetes/pull/57613)
+  against a component's command-line interface, which is required for components to maintain
+  explicit control over their command-line interface.
+
+Rather than require other components to copy and customize the Kubelet's relatively complex
+solutions to these problems, we should put the common functionality in a library, kflag, to reduce
+the burden on component owners.
+
+### Goals
+
+Provide core Kubernetes components with common building blocks that simplify the following:
+* Application of parsed flag values to arbitrary structs (e.g. config structs).
+* Selective inclusion of globally-registered flags in third-party code.
+
+### Non-Goals
+
+* This KEP is not concerned with fixing various problems with Cobra, such as third-party flags
+  leaking into the default usage or help text. We may write a future KEP to cover these issues and
+  command building in general.
+* This KEP is not concerned with the structure of flag and configuration types or their composition
+  into aggregates. We may write a future KEP to cover this.
+* This KEP is not concerned with loading and parsing config files, or resolving references contained
+  in those files (e.g. relative paths).
+* This KEP is not concerned with dynamically configuring components, or otherwise managing
+  component lifecycle.
+* This KEP is not concerned with flag or config validation.
+
+## Proposal
+
+This proposal recommends writing a wrapper package around the
+[pflag](https://godoc.org/github.com/spf13/pflag) library.
+This wrapper package is called _kflag_ in this proposal.
```

Comment:
important detail: will this be `k8s.io/kflag`, `k8s.io/utils/kflag` or `k8s.io/component-base/cli/klalg`? We don't have any dependencies to other kube libraries as far as I see. So we could go with one of the first two.

Reply:
I was thinking `k8s.io/component-base/cli/kflag`, at least while we are implementing the early iterations. Giving it it's own repo could make sense as well, given `k8s.io/klog`.

Reply:
we can start in experimental/cli/kflag and lateron move it to its own repo. I don't think it is a good idea to start in cli/kflag if we plan to move it anyway (and IMO we should at some point, when we have stable enough code).

Reply:
sgtm